daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala

```diff
@@ -0,0 +1,31 @@
+// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.daml.lf.data
+
+import scala.util.matching.Regex
+
+sealed abstract class MatchingStringModule {
+  type T <: String
+
+  def fromString(s: String): Either[String, T]
+
+  @throws[IllegalArgumentException]
+  def assertFromString(s: String): T =
+    fromString(s).fold(e => throw new IllegalArgumentException(e), identity)
+
+  def unapply(x: T): Option[String] = Some(x)
```

Comment:
Totality checking actually understands the implication of this:

```suggestion
  def unapply(x: T): Some[String] = Some(x)
```

Reply:
fixed